kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -481,18 +481,36 @@ protected void handleInactive(KafkaHeaderAndRequest kafkaHeaderAndRequest,
                 }
             } else if (e instanceof PulsarAdminException.NotFoundException) {
                 if (allowAutoTopicCreation) {
-                    log.info("[{}] Topic {} doesn't exist, auto create it with {} partitions",
-                            ctx.channel(), topicName, defaultNumPartitions);
-                    admin.topics().createPartitionedTopicAsync(topicName, defaultNumPartitions)
-                            .whenComplete((__, createException) -> {
-                                if (createException == null) {
-                                    future.complete(defaultNumPartitions);
-                                } else {
-                                    log.warn("[{}] Failed to create partitioned topic {}: {}",
-                                            ctx.channel(), topicName, createException.getMessage());
-                                    future.complete(TopicAndMetadata.INVALID_PARTITIONS);
-                                }
-                            });
+                    String namespace = TopicName.get(topicName).getNamespace();
+                    admin.namespaces().getPoliciesAsync(namespace).whenComplete((policies, err) -> {
+                        if (err != null || policies == null) {
+                            log.error("[{}] Cannot get policies for namespace {}", ctx.channel(), namespace, err);
+                            future.complete(TopicAndMetadata.INVALID_PARTITIONS);
+                        } else {
+                            boolean allowed = kafkaConfig.isAllowAutoTopicCreation();
+                            if (policies.autoTopicCreationOverride != null) {
+                                allowed = policies.autoTopicCreationOverride.isAllowAutoTopicCreation();
+                            }
+                            if (!allowed) {
+                                log.error("[{}] Automatic topic creation is not allowed on namespace {}",
+                                        ctx.channel(), namespace);
+                                future.complete(TopicAndMetadata.AUTHORIZATION_FAILURE);
```

Comment:
```suggestion
                                future.complete(TopicAndMetadata.INVALID_PARTITIONS);
```

Reply:
In addition, we should not return an authorization error to Kafka client even if this error is caused by a permission validation failure. Kafka client should receive the authorization error only if the client has sent some authorization info.

Reply:
makes sense, I will update the patch